### ISSUE TITLE
[bug] NamedTuple of same types are not compatible

### ISSUE THREAD

**Issue by alexeypetrushin** [crystal-lang/crystal#8652]
```
alias Data = NamedTuple(
  title: String?
)

data : Data = { title: nil }     # Explicitly specifying that 
                                 # it has ` : Data` type 

proc = ->(data : Data){ p data } 
proc.call data                   # And yet it says it's not 
                                 # of `Data` type
```

**Comment by straight-shoota** [comment]
Context: https://stackoverflow.com/questions/59598116/why-namedtuple-of-same-types-are-not-compatible

As explained in my answer on SO, the explanation about the assignment is incorrect. The type restriction for the variable doesn't specify the type of the literal, which is only `NamedTuple(title: Nil)`, not `NamedTuple(title: String | Nil)`.

But the actual type is accepted because it's covariant. It should also be valid for the proc's type restriction. That's the bug.

**Comment by asterite** [comment]
I think the bug is that if you say `data : Data`, `data` should be of type `Data`, not a type restriction. Type restrictions exist in method arguments. Anything else is a type declaration, like when you do `@data : Data`.

So the bug is that `data : Data` acts as a type restriction and not as a type declaration.

**Comment by straight-shoota** [comment]
If `data : Date` strictly makes `data` be `Data`, what happens with subtypes of `Data`?

**Comment by asterite** [comment]
What do you mean?

**Comment by straight-shoota** [comment]
It seems the compiler considers `NamedTuple(title: Nil)` as a compatible subtype of `NamedTuple(title: String | Nil)`. Can type `NamedTuple(title: Nil)` be assigned to a variable declared `NamedTuple(title: String | Nil)`?

**Comment by asterite** [comment]
Yes. But it can't be passed to a proc accepting that type because a function pointer accepts a specific type, there can't be an implicit cast.

**Comment by RX14** [comment]
`data : NamedTuple(title: String?) = { title: nil }` should either fail, or change the type inference on the right hand side. The proc is irrelevant.
